### PR TITLE
Streaming: replace npmlog with pino & pino-http

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This implements #27387. 

**Advice for reviewers:** I'd recommend going commit by commit in the files changed tab, as each commit tries to focus in on specific changes (though the change from `verifyClient` to `server.on("upgrade")` is bundled with the logger change)

In order to properly initialize the logging, I had to rework the WebSocket Server attachment to the HTTP server to use the upgrade event, which allows for making a decision around authentication before even attempting to upgrade the connection, which is preferrable to `verifyClient` whose usage is now discouraged. 

By using the upgrade event and `ws.handleUpgrade` method allows us to pass through a logger to the `wss.on("connection")` event/handler, which can be augmented with logging details specific to websockets.

The log redaction and sanitizer is a little complicated, but essentially prevents access tokens from accidentally ending up in server logs.

The `request.id` is now managed by the logger, and we don't need `request.remoteAddress` as the request logger correctly captures the information about the request for logging, so we're able to remove both of those middlewares. Additionally, I replaced the custom CORS middleware with the standard and widely used `cors` package, which is well tested.

Bundled in is the start of some code cleanup, moving things out to separate files where appropriate (since these things where making it much more difficult to read the code whilst I changed the connection upgrade handling).

In development, if you want, you can pipe to pino-pretty for nicer looking logs, which is available via a development dependency:

```sh
$ yarn workspace @mastodon/streaming start | pino-pretty
```

Screenshot 2023-11-12 at 9 17 57 pm

### CODE REVIEW

**Comment by ThisIsMissEm** [comment]
@renchap  [9b24627](https://github.com/mastodon/mastodon/pull/27828/commits/9b2462731af4074d30440d008c7881d251b8b851) is what I was talking about over on #24702, it didn't directly break anything when linting ran, but did cause issues; This will also act as a starting point for eventually being able to migrate to ESM, without affecting the reset of the codebase.

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:102]: I am tempted to move this into a generic `setupLogging` function that takes in the environment, and returns a `logger` and `httpLogger` back and also exposes the `sanitizeRequestForLogging` (`sanitizeReqLog`) method.

Would appreciate thoughts on if this makes sense as a change?

**Comment by mjankowski** [comment]
This is definitely easier to process one commit at a time than all together, but even so, it seems like at least:

- The eslint/tsconfig changes
- Extracting the metrics to a separate file
- Move isTruthy to utils
- Swap out custom impl for `cors` package

...could all be stand-alone PRs (they all seem like improvements), and may make it easier to review/merge the narrower stated purpose here of the logger change?

**Comment by ThisIsMissEm** [comment]
@mjankowski I've tried in the past to split changes down, but due to the project merge cadence I had to abandon those PRs which were all stacked and a nightmare to maintain / rebase.

I'm trying to avoid that here.

**Comment by ThisIsMissEm** [comment]
Extracted the CORS change to https://github.com/mastodon/mastodon/pull/28523/ & rebased this PR to separate the server initialization changes from the logging changes as best as possible.

File: streaming/logging.js
Comment (ThisIsMissEm) [inline streaming/logging.js:33]: @ClearlyClaire could you confirm that access tokens conform to this regex? I believe they should, but I'm not 100% sure.

**Comment by renchap** [comment]
Now that the CORS refactor is merged, could you extract your websocket refactor commit into it's own PR?

Then I think we will be able to review & merge this :)

Thanks a lot for this work!

**Comment by ThisIsMissEm** [comment]
> Now that the CORS refactor is merged, could you extract your websocket refactor commit into it's own PR?
> 
> Then I think we will be able to review & merge this :)
> 
> Thanks a lot for this work!

With all due respect, this is the fourth time I've done the work in this branch & tried to get it merged. I've had to abandon 3 other attempts. My time, like yours is extremely limited. I have tried to make the changes reviewable commit by commit: https://github.com/mastodon/mastodon/pull/27828/commits/6533a523d37e5cf5afb7ec012409f141f2064e33 & https://github.com/mastodon/mastodon/pull/27828/commits/93351e1c6c865c281687dfd190720246de58b8bb

I do not have time to continuously rebase this work, nor do I have the energy as I'm currently reasonably unwell. You either want this change and for the streaming server to improve, or you don't. I know your resources are thin, but mine are likely thinner. I've been trying to get this work merged for almost a year at this point.

**Comment by ThisIsMissEm** [comment]
There is a follow-on related to this pull request here: https://github.com/mastodon/mastodon/pull/28632

File: streaming/logging.js
Comment (ClearlyClaire) [reply streaming/logging.js:33]: As far as I understand Devise's code, this uses `SecureRandom.urlsafe_base64` without padding, so this looks accurate!

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:41]: These `@ts-ignore`s are due to a typing issue with the `err` parameter here.

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:74]: I'd ideally like to remove this mess, and instead handle the `parseJSON` method as a Result object `{ error, value }`, which would allow us to pass the error back to the call-site for logging, rather than performing the logging in this method.

I'm unsure if I should make this change now or not.

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:219]: This is only used for the responseTime logging, which I'm not sure if we really need here, so we could remove this and line 267.

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:279]: The logger child that we're creating here is explicitly a child of the root logger and not `request.log` as the former would include all the request properties with every websocket log message, which would be excessive. Instead we're just passing through `request.id` for correlation purposes.

I could perhaps rename `websocketLogger` to `websocketRequestLogger` and have a `createWebsocketLogger(request)` here.

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:287]: For now I'm adding a third argument to `onConnection`, but in the future, we may be able to attach the "logger" directly to the `websocket` (`ws`)

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:318]: These were due to unused parameters, we may not need those `@ts-ignore`s anymore.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire I'm just refactoring a bit of this into a separate PR which tackles the type issues with Redis & Postgresql (I'm going to extract these things to separate files since the fixes ended up being kind of messy)

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:447]: This is because the types for `pg` don't declare it as a union on `Pool.connect` of `(err: Error, client: never, done: (release?: any) => void) | (err: never, client: PoolClient, done: (release?: any) => void)` which would allow discrimination of the types, so instead you get an error of "client may be undefined" for `client.query`

We could alternatively use async/await here, which would side-step that issue of `Pool.connect`'s types being incorrect since: `const client = await pgPool.connect()` is correctly typed as `Promise<PoolClient>;`

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:459]: This is being refactored separately, as I've a branch that introduces proper Error objects for the various errors we throw in the code.

File: streaming/index.js
Comment (ThisIsMissEm) [inline streaming/index.js:590]: I have a separate commit that completely refactors this code, so for now I've just added a minor guard to ensure messages from ruby over redis to streaming are well-formed.

File: streaming/index.js
Comment (ThisIsMissEm) [reply streaming/index.js:419]: I could also change this to 
```suggestion
   * @param {http.IncomingMessage & ResolvedAccount} req
```

As I have done so elsewhere.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire @mjankowski okay, I think this should be good to go; I've moved the postgresql and redis type issues into a separate PR (still to open / finish, because I want to move those out of `index.js`)

File: streaming/index.js
Comment (ClearlyClaire) [reply streaming/index.js:210]: I don't feel strongly either way. I think this is not useful in most of cases, but it could be in some?

**Comment by ClearlyClaire** [comment]
Apart from the minor naming issues, this looks fine to me, but I'd like another set of eyes on this.

Also, this is a (fairly minor) breaking change that needs to be documented.

File: streaming/index.js
Comment (ThisIsMissEm) [reply streaming/index.js]: Have updated this to `attachWebsocketHttpLogger` and created a `createWebsocketLogger` for the actual logger below in the handleUpgrade.

**Comment by ThisIsMissEm** [comment]
> Also, this is a (fairly minor) breaking change that needs to be documented.

@ClearlyClaire currently the documentation site has absolutely zero mention of logging, and only a little bit on monitoring Mastodon. I did add some documentation about streaming's metrics to the "scaling up" docs, but I don't see a good place for the notice that this is a breaking change in how the streaming server produces logs.

By the looks the changelog should be updated in the release branch.

**Comment by ClearlyClaire** [comment]
> > Also, this is a (fairly minor) breaking change that needs to be documented.
> 
> @ClearlyClaire currently the documentation site has absolutely zero mention of logging, and only a little bit on monitoring Mastodon. I did add some documentation about streaming's metrics to the "scaling up" docs, but I don't see a good place for the notice that this is a breaking change in how the streaming server produces logs.
> 
> By the looks the changelog should be updated in the release branch.

Yes, I meant that as a note to add something in the release notes when the time comes, not as something to do before the PR gets merged.

**Review by ClearlyClaire — APPROVED** [review]

**Review by renchap — APPROVED** [review]
Looks good to me. Quite a hard change to test entirely, but it makes the code cleaner and hopefully provides more useful logs.

**Pull request merged**
